src/cpp/dds/domain/DomainParticipantImpl.cpp

```diff
@@ -135,38 +135,58 @@ DomainParticipantImpl::~DomainParticipantImpl()
 }
 
 
-bool DomainParticipantImpl::delete_publisher(
+ReturnCode_t DomainParticipantImpl::delete_publisher(
         Publisher* pub)
 {
+    if(participant_ != pub->get_participant())
+    {
+        return ReturnCode_t::RETCODE_PRECONDITION_NOT_MET;
+    }
     std::lock_guard<std::mutex> lock(mtx_pubs_);
     auto pit = publishers_.find(pub);
 
     if (pit != publishers_.end() && pub->get_instance_handle() == pit->second->get_instance_handle())
     {
+        std::vector<DataWriter*> writers;
+        pub->get_datawriters(writers);
+        if(!writers.empty())
```

Comment:
It will be a good optimization to add a `bool has_datawriters() const` method or something similar to the `Publisher` to avoid the copy.

```suggestion
        if (!writers.empty())
```

Reply:
Solved in PR #604